### DISCUSSION TITLE
Bootstrap server issue: Accessing Kafka cluster in a client Java Application

### DISCUSSION THREAD

**Discussion by talhagithub02** [strimzi/strimzi-kafka-operator#9641] [Q&A]
Hi every1, 
I need your assistance as I am stuck in to this: 
I am running Skrimzi on Windows pc with Docker desktop Kubernetes on my local pc. I am able to create cluster (configuration image attached) and it is in healthy state , I can send data to a topic and see in producer, but when I am trying to access that cluster using Java , it is not recognizing the bootstrap server (Error: No resolvable bootstrap urls given in bootstrap.servers). Here is what I have tried: 

1. using this command "kubectl get services -n kafka" i got the name of the bootstrap server "my-cluster-kafka-bootstrap" but when i am passing this in to Java 

`String bootstrapServers = "my-cluster-kafka-external-bootstrap:9094";

        // Create AdminClient properties
        java.util.Properties properties = new java.util.Properties();
        properties.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);`

I am receiving the unresolvable bootstrap error.

2. I tried adjusting config file to include 

`listeners:
      - name: external
        port: 9094
        type: nodeport
        tls: true` 
        
and tried using docker ip "192.168.65.3" with this Port , still same issue.

I tired many other ways but bounding this to only these two, 

One thing to mention is I came across this recent consumer blog: https://strimzi.io/blog/2023/11/09/kafka-consumer-client-essentials/

which mentions to get bootstrap server from this command: 
`kubectl get kafka <kafka_cluster_name> -o=jsonpath='{.status.listeners[*].bootstrapServers}{"\n"}' ` but this is returning me Nothing.

 Any output you need from myside, I can provide, I need to access the cluster inside Java application on my local. 

![Screenshot_11](https://github.com/strimzi/strimzi-kafka-operator/assets/132895376/0cca3b62-1332-414b-ae51-fd95d74fe3d1)

**Comment by scholzj** [comment]
I do not have Windows, so it is hard to say how exactly it works there. Node ports typically work for example on Minikube run with VM driver (as a virtual machine). But it typically does not work with Kubernetes running inside of a Docker container inside a VM as the _Kubernetes worker node_ is in a different network. You can check this blog post series to better understand how it works: https://strimzi.io/blog/2019/04/17/accessing-kafka-part-1/. You could also try this project I created: https://github.com/scholzj/kekspose (but as I said, I do not have Windows, so never had a chance to test it there)
